### PR TITLE
[Range slider] test dual thumb key press

### PR DESCRIPTION
**Pull request by danrosenthal**
Addresses #841

|before|after|
|-|-|
|screen shot 2019-01-17 at 11 51 13 am|screen shot 2019-01-17 at 11 49 19 am|

### CODE REVIEW

File: src/components/RangeSlider/components/DualThumb/DualThumb.tsx [hunk 1]
```diff
@@ -361,7 +361,7 @@ export default class DualThumb extends React.Component<Props, State> {
   @autobind
   private incrementValueUpper() {
     this.setValue(
-      [this.state.value[1] + this.props.step, this.state.value[0]],
+      [this.state.value[0], this.state.value[1] + this.props.step],
```
Comment (danrosenthal) [inline src/components/RangeSlider/components/DualThumb/DualThumb.tsx]: We had this argument flip-flopped

File: src/components/RangeSlider/components/DualThumb/DualThumb.tsx [hunk 2]
```diff
@@ -206,7 +206,6 @@ export default class DualThumb extends React.Component<Props, State> {
               />
               <button
                 id={idLower}
-                testID="thumbLower"
```
Comment (solonaarmstrong) [inline src/components/RangeSlider/components/DualThumb/DualThumb.tsx]: 🎉

File: src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx [hunk 3]
```diff
@@ -390,9 +542,17 @@ describe('<DualThumb />', () => {
...
   });
 });
+
+function findThumbLower(containerComponent: ReactWrapper): ReactWrapper {
+  return containerComponent.find('button').first();
+}
+
+function findThumbUpper(containerComponent: ReactWrapper): ReactWrapper {
+  return containerComponent.find('button').last();
+}
```
Comment (solonaarmstrong) [inline src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx]: ❤️

File: src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx [hunk 4]
```diff
@@ -301,14 +303,164 @@ describe('<DualThumb />', () => {
...
+      simulateKeyDown(findThumbLower(dualThumb), Key.DownArrow);
+
+      expect(onChangeSpy).toHaveBeenCalledWith([9, 40], mockProps.id);
+    });
+
+    function simulateKeyDown(component: ReactWrapper, keyCode: Key) {
+      component.simulate('keyDown', {
+        keyCode,
+        preventDefault: noop,
+        stopPropagation: noop,
```
Comment (solonaarmstrong) [inline src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx]: I didn't realize these were necessary in a simulate. TIL

File: src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx [hunk 5]
```diff
@@ -301,14 +303,164 @@ describe('<DualThumb />', () => {
...
+      const onChangeSpy = jest.fn();
+      const dualThumb = mountWithAppProvider(
+        <DualThumb {...mockProps} value={[10, 40]} onChange={onChangeSpy} />,
+      );
+      simulateKeyDown(findThumbLower(dualThumb), Key.RightArrow);
+
+      expect(onChangeSpy).toHaveBeenCalledWith([11, 40], mockProps.id);
+    });
+
+    it('does not increment the lower value on upper value key press', () => {
```
Comment (solonaarmstrong) [inline src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx]: 👌

File: src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx [hunk 6]
```diff
@@ -301,14 +303,164 @@ describe('<DualThumb />', () => {
...
+      const onChangeSpy = jest.fn();
+      const dualThumb = mountWithAppProvider(
+        <DualThumb {...mockProps} value={[10, 40]} onChange={onChangeSpy} />,
+      );
+      simulateKeyDown(findThumbUpper(dualThumb), Key.RightArrow);
+
+      expect(onChangeSpy).toHaveBeenCalledWith([10, 41], mockProps.id);
+    });
+
+    it('does not decrement the lower value on upper value key press', () => {
```
Comment (solonaarmstrong) [inline src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx]:
```suggestion
    it('does not decrement the lower value on upper thumb key press', () => {
```

Not attached, I just find it clearer.

File: src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx [hunk 7]
```diff
@@ -301,14 +303,164 @@ describe('<DualThumb />', () => {
     });
   });
 
+  describe('keyboard control', () => {
```
Comment (tmlayton) [inline src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx]:
```suggestion
  describe('keyboard interface', () => {
```

File: src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx [hunk 8]
```diff
@@ -301,14 +303,164 @@ describe('<DualThumb />', () => {
...
+      const onChangeSpy = jest.fn();
+      const dualThumb = mountWithAppProvider(
+        <DualThumb {...mockProps} value={[10, 40]} onChange={onChangeSpy} />,
+      );
+      simulateKeyDown(findThumbLower(dualThumb), Key.DownArrow);
+
+      expect(onChangeSpy).toHaveBeenCalledWith([9, 40], mockProps.id);
+    });
+
+    it('does not increment the upper value on lower value key press', () => {
```
Comment (tmlayton) [inline src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx]:
```suggestion
    it('increment the lower value on up arrow press', () => {
```

File: src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx [hunk 9]
```diff
@@ -301,14 +303,164 @@ describe('<DualThumb />', () => {
...
+      const onChangeSpy = jest.fn();
+      const dualThumb = mountWithAppProvider(
+        <DualThumb {...mockProps} value={[10, 40]} onChange={onChangeSpy} />,
+      );
+      simulateKeyDown(findThumbLower(dualThumb), Key.UpArrow);
+
+      expect(onChangeSpy).toHaveBeenCalledWith([11, 40], mockProps.id);
+    });
+
+    it('does not decrement the upper value on lower value key press', () => {
```
Comment (tmlayton) [inline src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx]:
```suggestion
    it('decrement the lower value on down arrow press', () => {
```

**Comment by tmlayton** [comment]
Coverage is now ~95% according to Codecov – did a mock implementation of `addEventListener` and `removeEventListener` so that we can manually access and call the handlers with arbitrary event data.

File: src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx (hunk 4 above)
Comment (danrosenthal) [reply src/components/RangeSlider/components/DualThumb/tests/DualThumb.test.tsx]: Only necessary when the handler uses them. It's necessary to mock any property on the event object that we access in the handler.

**Pull request merged**